### PR TITLE
Enable partial data propagation for Resize's sizes

### PR DESCRIPTION
**Pull request by jcwchen**
### Description
Enable partial data propagation for Resize's sizes

### Motivation and Context
Current Resize's size cannot get symbolic shape from ONNX's data propagation. Closes https://github.com/onnx/onnx/issues/4580.

### CODE REVIEW

**Comment by jcwchen** [comment]
cc @justinchuby I have tested this PR with your provided model and it can infer Resize's shape as [1, 1, 2, 2].

**Comment by justinchuby** [comment]
Thank you!

**Comment by gramalingam** [comment]
Thanks Jacky for this. I wonder, however, if we can use [getInputShape](https://github.com/onnx/onnx/blob/a525f98d5ed31660b629bab90c680a39658a5c08/onnx/defs/shape_inference.cc#L428) ... I think it was abstracted as a function for this purpose.

**Comment by jcwchen** [comment]
> Thanks Jacky for this. I wonder, however, if we can use [getShapeInput](https://github.com/onnx/onnx/blob/a525f98d5ed31660b629bab90c680a39658a5c08/onnx/defs/shape_inference.cc#L428) ... I think it was abstracted as a function for this purpose.

Good catch. Yes, I should use getShapeInput instead. Updated. Please review it. Thanks!

File: onnx/defs/tensor/utils.cc [hunk 1]
```diff
@@ -157,54 +165,53 @@ void resizeShapeInferenceVersioned(InferenceContext& ctx, int opset_version) {
...
+      }
+    } else {
+      // sizes_data contains scales for all axes
+      if (sizes_data.size() != rank_x) {
+        fail_shape_inference(
+            "Number of elements of input 'sizes' (",
+            sizes_data.size(),
+            ") must be same as rank of input 'X' (",
+            rank_x,
+            ").");
```
Comment (jcwchen) [inline onnx/defs/tensor/utils.cc:194]: Actually I was confused about this check. I saw the document also mentioned this, but existing test models like `test_resize_tf_crop_and_resize_axes_2_3` and `test_resize_tf_crop_and_resize_axes_3_2` do not follow this. Take `test_resize_tf_crop_and_resize_axes_2_3` as an example, the model is (input rank 4 != sizes 2):

image

@gramalingam do you think they are invalid ONNX models?

Comment (gramalingam) [reply onnx/defs/tensor/utils.cc:194]: Hi, I think this may be a bug introduced when the `axes` attribute was added to Resize. If the `axes` attribute is specified, we need to account for it (and the PR that added the attribute may have failed to account for it both in code and the documentation).

Comment (gramalingam) [reply onnx/defs/tensor/utils.cc:194]: This PR: https://github.com/onnx/onnx/pull/4126

Comment (gramalingam) [reply onnx/defs/tensor/utils.cc:194]: Good catch!

File: onnx/defs/tensor/utils.cc [hunk 2]
```diff
@@ -96,20 +96,28 @@ void resizeShapeInferenceVersioned(InferenceContext& ctx, int opset_version) {
   bool hasSizesInput = ctx.hasInput(3);
 
   const TensorProto* scales = 2 < ctx.getNumInputs() ? ctx.getInputData(2) : nullptr;
-  const TensorProto* sizes = 3 < ctx.getNumInputs() ? ctx.getInputData(3) : nullptr;
+  std::vector<int64_t> sizes_data;
+  if (3 < ctx.getNumInputs()) {
+    bool found_sizes = false;
+    const auto sizes_shape = getShapeInput(ctx, 3, found_sizes);
+    // If sizes is an empty shape, assume it's not provided
```
Comment (gramalingam) [inline onnx/defs/tensor/utils.cc:103]: Can you clarify why? Specifically: is my understanding correct that this is an error-situation? Or, is this supposed to be a valid situation? (I guess it could be valid if `axes` is empty ... except for that edge-case.)

Comment (gramalingam) [reply onnx/defs/tensor/utils.cc:103]: Oh ... I remember, is this the old test-case fix where empty-tensor was used (before input was made optional)?

Comment (jcwchen) [reply onnx/defs/tensor/utils.cc:103]: https://github.com/onnx/onnx/pull/4388 I think it just provides some flexibility to let ONNX handle old models.

>  I remember, is this the old test-case fix where empty-tensor was used (before input was made optional)?

Yes! Some existing models in ONNX Model Zoo have this kind of usage.

File: onnx/defs/tensor/utils.cc (hunk 1 above)
Comment (jcwchen) [reply onnx/defs/tensor/utils.cc:194]: Thank you for the helpful context! Sorry that I think I provided the wrong model above. Here is the right model for `test_resize_tf_crop_and_resize_axes_2_3`.

image


Based on the naming of this test model, does it mean that this model should have axis attribute?

Comment (gramalingam) [reply onnx/defs/tensor/utils.cc:194]: So, it is difficult to see what was in the old-version vs. new-version here ... the check is fine if axes attribute wasn't specified, right? The new-version seems to have this check only if axes attribute is not specified?

Comment (jcwchen) [reply onnx/defs/tensor/utils.cc:194]: IIUC, the only difference about this issue is the update from this PR enables "sizes" input to have the shape information even if the input tensor is unknown. Previously the tensor (sizes) will be nullptr anyway so it did not check the rank for input x and input sizes... IMO, a possible fix might be making these 2 models have attribute axes (2, 3) and axes (3, 2) and then they should pass the existing check. Does it look good to you?

Comment (gramalingam) [reply onnx/defs/tensor/utils.cc:194]: I agree the test-cases seem to be missing the axes attribute. We should add them. I guess you are saying that the checker wasn't catching the error previously, and we are able to catch it now (because of the improvement due to data-propagation)?

Comment (jcwchen) [reply onnx/defs/tensor/utils.cc:194]: > I guess you are saying that the checker wasn't catching the error previously, and we are able to catch it now (because of the improvement due to data-propagation)?

Yes. data-propagation will at least keep the shape information (the shape of "sizes" in this case) so it can do more check. I will update these models then. Thanks!

Comment (jcwchen) [reply onnx/defs/tensor/utils.cc:194]: I have updated these 2 models with attribute axes. Now both of them have passed onnx.shape_inference. PTAL.

**Review by gramalingam — APPROVED** [review]
LGTM, thanks!

**Comment by wine99** [comment]
If I understand correctly, all the nodes in this graph support data propagation. However, running

```
inferred_model = shape_inference.infer_shapes(model, strict_mode=True, data_prop=True)
```
still leaves me with four unknown dimensions. Could you help me identify the problem here? I am using the python package onnx 1.17.0

image
